providers/implementations/kdfs/argon2.c

```diff
@@ -0,0 +1,1488 @@
+{
+    KDF_ARGON2 *ctx;
+
+    ctx = OPENSSL_zalloc(sizeof(*ctx));
+    if (ctx == NULL) {
+        ERR_raise(ERR_LIB_PROV, ERR_R_MALLOC_FAILURE);
+        return NULL;
+    }
+
+    ctx->provctx = provctx;
+
+    kdf_argon2_init(ctx, ARGON2_ID);
+    return ctx;
+}
+
+static void kdf_argon2_free(void *vctx)
+{
+    ARGON2_CTX *ctx;
+    KDF_ARGON2 *ctx_wrap;
+
+    ctx_wrap = (KDF_ARGON2 *) vctx;
+    ctx = &ctx_wrap->ctx;
+    if (ctx->out != NULL)
+        OPENSSL_clear_free(ctx->out, ctx->outlen);
+
+    if (ctx->pwd != NULL)
+        OPENSSL_clear_free(ctx->pwd, ctx->pwdlen);
+
+    if (ctx->salt != NULL)
+        OPENSSL_clear_free(ctx->salt, ctx->saltlen);
+
+    if (ctx->secret != NULL)
+        OPENSSL_clear_free(ctx->secret, ctx->secretlen);
+
+    if (ctx->ad != NULL)
+        OPENSSL_clear_free(ctx->ad, ctx->adlen);
+
+    memset(ctx, 0, sizeof(*ctx));
+    OPENSSL_free(ctx_wrap);
+}
+
+static int kdf_argon2_derive(void *vctx, unsigned char *out, size_t outlen)
+{
+    ARGON2_CTX *ctx;
+    KDF_ARGON2 *ctx_wrap;
+    uint32_t memory_blocks, segment_length;
+
+    ctx_wrap = (KDF_ARGON2 *)vctx;
+    ctx = &ctx_wrap->ctx;
+
+    if (ctx->pwd == NULL || ctx->pwdlen == 0) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_MISSING_PASS);
+        return 0;
+    }
+
+    if (ctx->salt == NULL || ctx->saltlen == 0) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_MISSING_SALT);
+        return 0;
+    }
+
+    if (outlen > UINT32_MAX)
+        return 0;
+
+    if (outlen == 0)
+        return 0;
+
+    if (outlen != ctx->outlen)
+        kdf_argon2_ctx_set_digest_length(ctx, (uint32_t) outlen);
+
+    if (ctx->type != ARGON2_D)
+        ctx->out = OPENSSL_secure_zalloc(ctx->outlen+1);
+    else
+        ctx->out = OPENSSL_zalloc(ctx->outlen+1);
+
+    if (validate_inputs(ctx) != 1)
+        goto fail;
+
+    memory_blocks = ctx->m_cost;
+    if (memory_blocks < 2 * ARGON2_SYNC_POINTS * ctx->lanes)
+        memory_blocks = 2 * ARGON2_SYNC_POINTS * ctx->lanes;
+
+    /* Ensure that all segments have equal length */
+    segment_length = memory_blocks / (ctx->lanes * ARGON2_SYNC_POINTS);
+    memory_blocks = segment_length * (ctx->lanes * ARGON2_SYNC_POINTS);
+
+    ctx->memory = NULL;
+    ctx->memory_blocks = memory_blocks;
+    ctx->segment_length = segment_length;
+    ctx->passes = ctx->t_cost;
+    ctx->lane_length = segment_length * ARGON2_SYNC_POINTS;
+
+    if (ctx->threads > ctx->lanes)
+        ctx->threads = ctx->lanes;
+
+    if (initialize(ctx) != 1)
+        goto fail;
+
+    if (fill_memory_blocks(ctx) != 1)
+        goto fail;
+
+    finalize(ctx);
+    memcpy(out, ctx->out, outlen);
+
+    return 1;
+
+ fail:
+    if (ctx->type != ARGON2_D)
+        OPENSSL_secure_clear_free(ctx->out, ctx->outlen+1);
+    else
+        OPENSSL_clear_free(ctx->out, ctx->outlen+1);
+    ctx->out = NULL;
+    return 0;
+}
+
+static void kdf_argon2_reset(void *vctx)
+{
+    KDF_ARGON2 *ctx_wrap;
+    ARGON2_CTX *ctx;
+    ARGON2_TYPE type;
+
```

Comment:
Like the other KDF reset functions there is a bug where the 
ctx->provctx is being NULLED here..
It needs to be saved at the start and restored between the memset and init()